compiler/z/codegen/S390Peephole.cpp

```diff
@@ -1349,29 +1349,89 @@ TR_S390PostRAPeephole::ConditionalBranchReduction(TR::InstOpCode::Mnemonic branc
-   TR::InstOpCode::Mnemonic nextOpCode = next->getOpCodeValue();
-   if ((curOpCode == TR::InstOpCode::CR || curOpCode == TR::InstOpCode::CGR || curOpCode == TR::InstOpCode::CGFR)
-      && (nextOpCode == TR::InstOpCode::BRC || nextOpCode == TR::InstOpCode::BRCL))
-      {
-      printf("Finding CR + BRC\n");
-      printf("method=%s\n", comp()->signature());
+   bool branchTakenPerformReduction = false;
+   bool fallThroughPerformReduction = false;
+   TR::Instruction *clrInstruction = _cursor;
+   TR::Instruction *brcInstruction = _cursor->getNext();
+   TR::Instruction *clrjInstruction = NULL;
+   TR::LabelSymbol *labelSymbol = NULL;
+
+   if (clrInstruction->getOpCodeValue() == TR::InstOpCode::CLR
+       && brcInstruction->getOpCodeValue() == TR::InstOpCode::BRC)
+      {
+      /* if branch target is a snippet we'll go ahead with the
+       * reduction. Otherwise, we look at next 5 (or until TR::BBEnd
+       * depending on whichever comes first) to check if any
+       * instruction reads or sets the condition code.
+       * We'll perform the reduction if an instruction sets
+       * the condition code or, if none of the instructions read the
+       * condition code set by clrInstruction
+       */
+      labelSymbol = brcInstruction->getLabelSymbol();
+      if (labelSymbol->getSnippet())
+         {
+         branchTakenPerformReduction = true;
+         }
+      else
+         {
+            // check branch case for condition code usage
+            int branchTakenInstIndex = 0;
+            TR::Instruction* targetInstruction = labelSymbol->getInstruction();
+            while (branchTakenInstIndex < 5
+               && targetInstruction
+               && (branchTakenInstIndex > 0 && !targetInstruction->getPrev()->getOpCode().setsCC())
+               && TR::BBEnd != targetInstruction->getNode()->getOpCodeValue())
+            {
+               branchTakenPerformReduction = !targetInstruction->getOpCode().readsCC();
+               ++branchTakenInstIndex;
+               targetInstruction = targetInstruction->getNext();
+            }
+
+         }
+      // check fall through case for condition code usage
+      int fallThroughInstIndex = 0;
+      TR::Instruction* fallThroughInstruction = brcInstruction->getNext();
+      while (fallThroughInstIndex < 5
+         && fallThroughInstruction
+         && !fallThroughInstruction->getPrev()->getOpCode().setsCC()
+         && TR::BBEnd != fallThroughInstruction->getNode()->getOpCodeValue())
+         {
+            fallThroughPerformReduction = !fallThroughInstruction->getOpCode().readsCC();
+            ++fallThroughInstIndex;
+            fallThroughInstruction = fallThroughInstruction->getNext();
+         }
+
+      if (fallThroughPerformReduction && branchTakenPerformReduction)
+         {
+         TR_ASSERT(clrInstruction->getNumRegisterOperands() == 2, "Number of register operands was not 2: %d\n",clrInstruction->getNumRegisterOperands());
```

Comment:
Prefer to use `TR_ASSERT_FATAL` if the code following the assertion will result in potentially functionally incorrect code which it will in this case. `TR_ASSERT` is turned off in production builds. Also a space needed in between arguments (`as not 2: %d\n",clrInstru`).